pkg/occlient/volumes.go

```diff
@@ -120,31 +120,84 @@ func (c *Client) getVolumeNamesFromPVC(pvc string, dc *appsv1.DeploymentConfig)
 
 // removeVolumeFromDC removes the volume from the given Deployment Config and
 // returns true. If the given volume is not found, it returns false.
-func removeVolumeFromDC(vol string, dc *appsv1.DeploymentConfig) bool {
+func removeVolumeFromDC(vol string, dc *appsv1.DeploymentConfig) error {
+
+	// Error out immediately if there are zero volumes to begin with
+	if len(dc.Spec.Template.Spec.Volumes) == 0 {
+		return errors.New("there are *no* volumes in this DeploymentConfig to remove")
+	}
+
 	found := false
-	for i, volume := range dc.Spec.Template.Spec.Volumes {
-		if volume.Name == vol {
-			found = true
-			dc.Spec.Template.Spec.Volumes = append(dc.Spec.Template.Spec.Volumes[:i], dc.Spec.Template.Spec.Volumes[i+1:]...)
+
+	// If for some reason there is only one volume, let's slice the array to zero length
```

Comment:
The reason is because you are editing the list that you are iterating over and there is no `break` statement after we found a volume.
e.g.
the volumes list `dc.Spec.Template.Spec.Volumes ` initially has 2 volumes. then when a volume is found with the matching name - `dc.Spec.Template.Spec.Volumes` shinks by 1 length but we don't break the loop there so if the next volume also has the same name 
```
dc.Spec.Template.Spec.Volumes = append(dc.Spec.Template.Spec.Volumes[:i], dc.Spec.Template.Spec.Volumes[i+1:]...)
```
will fail.

So I would suggest maybe consider just adding a `break` statement in the older code and see if that works @cdrage

Reply:
and this issue is present here
https://github.com/openshift/odo/blob/5cb97bcd7dbfbc243e8405857fadea268c4e3db2/pkg/occlient/volumes.go#L171
https://github.com/openshift/odo/blob/5cb97bcd7dbfbc243e8405857fadea268c4e3db2/pkg/occlient/volumes.go#L143
and here
https://github.com/openshift/odo/blob/5cb97bcd7dbfbc243e8405857fadea268c4e3db2/pkg/occlient/volumes.go#L128
just from the first glance

Reply:
Yup. That's intentional. That is where the issue was.

The comment that "if for some reason there is only 1 volume" is that sometimes when editing the deployment config there is actually only *one* volume available rather than two (s2i bootstrap image & the volume for source code).

So that's the reason for the if statement. Although you are right, there should be a break in the for loop.